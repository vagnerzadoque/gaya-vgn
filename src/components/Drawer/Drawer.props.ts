export type DrawerSize = 'small' | 'medium' | 'large'
export type DrawerPosition = 'left' | 'right' | 'top' | 'bottom'
export interface DrawerProps {
  /**
   * The content of Drawer
   */
  children: React.ReactNode
  /**
   * Override or extend the styles applied to the component
   */
  className?: string
  /**
   * Whether the Drawer is opened or not
   */
  open: boolean
  /**
   * Whether the Drawer is opened use to close in overlay
   */
  handleClick?: () => void
  /**
    * The Drawer's width size
    * @default small
  */
  size?: DrawerSize
  /**
   * Optional ID for testing
   */
  testID?: string
  /**
   * Optional position prop
   */
  position?: DrawerPosition
}
