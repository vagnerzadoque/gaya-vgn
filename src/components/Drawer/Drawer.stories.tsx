/* eslint-disable max-statements-per-line */
/* eslint-disable max-len */
/* eslint-disable max-lines */
/* eslint-disable react/no-array-index-key */
import React, { useState } from 'react'
import type { Meta, StoryObj } from '@storybook/react'
import { fn } from '@storybook/test'
import { IconName } from '@naturacosmeticos/natds-icons'
import { createUseStyles } from 'react-jss'
import { Theme } from '@naturacosmeticos/natds-themes'
import { Drawer, DrawerProps } from '.'
import { IconButton } from '../IconButton'
import { Button } from '../Button'
import { Icon } from '../Icon'
import { Avatar } from '../Avatar'
import { Divider } from '../Divider'
import { ListItem } from '../ListItem'
import { Badge } from '../Badge'

const componentStatus = `
> A Drawer is a thin line that groups content in lists and layouts.

## Properties
| Property                      | Values                                |    Status           |
|---                            |                                    ---|                  ---|
| **variant(no prop)**          | Base                                  | ✅ Available        |
| **open**                      | true/false                            | ✅ Available        |
| **handleClick**               | function                              | ✅ Available        |
| **size**                      | small, medium, large                  | ✅ Available        |
| **position**                  | left, right, top, bottom              | ✅ Available        |

## Technical Usages Examples
`

const meta: Meta<DrawerProps> = {
  title: 'Components/Drawer',
  component: Drawer,
  parameters: {
    componentSubtitle: '',
    docs: { description: { component: componentStatus } }
  }
};

export default meta;
type Story = StoryObj<DrawerProps>;

export const Playground: Story = {
  render: (args) => {
    const [showDrawer, setShowDrawer] = useState(false)
    const {
      footer,
      header,
      headerButton,
      headerTitle,
      main,
      sectionItem
    } = styles()
    return (
      <div style={{
        overflow: 'hidden', maxWidth: '1280px', maxHeight: '980px'
      }}
      >
        <Button onClick={() => setShowDrawer(!showDrawer)}>Open Drawer</Button>
        <Drawer {...args} open={showDrawer} handleClick={() => setShowDrawer(!showDrawer)}>
          <div className={header}>
            <Avatar type="icon" />
            <h6 className={headerTitle}>Hello, Design System</h6>
            <div className={headerButton}>
              <IconButton
                onClick={() => setShowDrawer(!showDrawer)}
                ariaLabel="close drawer"
                IconComponent={<Icon name="outlined-navigation-close" />}
              />
            </div>
          </div>
          <Divider />
          <div className={main}>
            <ul style={{ padding: 0, margin: 0, width: '100%' }}>
              {items.body.map((item, id) => (
                <ListItem key={id} interaction="action">
                  <div style={{ display: 'flex', gap: 16, alignItems: 'center' }}>
                    <Icon name={item.icon as IconName} />
                    <span className={sectionItem}>
                      {item.name}
                      {item.new && <Badge variant="pulse" />}
                    </span>
                  </div>
                </ListItem>
              ))}
            </ul>
          </div>
          <div className={footer}>
            <Divider />
            <ul style={{ padding: 0, margin: 0, overflow: 'auto' }}>
              {items.footer.map((item, id) => (
                <ListItem key={id} interaction="action">
                  <div style={{ display: 'flex', gap: 16, alignItems: 'center' }}>
                    <Icon name={item.icon as IconName} />
                    <span className={sectionItem}>
                      {item.name}
                      {item.new && <Badge variant="pulse" />}
                    </span>
                  </div>
                </ListItem>
              ))}
            </ul>
          </div>
        </Drawer>
      </div>
    )
  }
}
export const Right: Story = {
  args: {
    position: "right"
  },
  render: (args) => {
    const [showDrawer, setShowDrawer] = useState(false)
    const {
      footer,
      header,
      headerButton,
      headerTitle,
      main,
      sectionItem
    } = styles()
    return (
      <div style={{
        overflow: 'hidden', width: '100%', height: '100%'
      }}
      >
        <Button onClick={() => setShowDrawer(!showDrawer)}>Open Drawer</Button>
        <Drawer {...args} open={showDrawer}>
          <div className={header}>
            <Avatar type="icon" />
            <h6 className={headerTitle}>Hello, Design System</h6>
            <div className={headerButton}>
              <IconButton
                onClick={() => setShowDrawer(!showDrawer)}
                ariaLabel="close drawer"
                IconComponent={<Icon name="outlined-navigation-close" />}
              />
            </div>
          </div>
          <Divider />
          <div className={main}>
            <ul style={{ padding: 0, margin: 0, width: '100%' }}>
              {items.body.map((item, id) => (
                <ListItem key={id} interaction="action">
                  <div style={{ display: 'flex', gap: 16, alignItems: 'center' }}>
                    <Icon name={item.icon as IconName} />
                    <span className={sectionItem}>
                      {item.name}
                      {item.new && <Badge variant="pulse" />}
                    </span>
                  </div>
                </ListItem>
              ))}
            </ul>
          </div>
          <div className={footer}>
            <Divider />
            <ul style={{ padding: 0, margin: 0, overflow: 'auto' }}>
              {items.footer.map((item, id) => (
                <ListItem key={id} interaction="action">
                  <div style={{ display: 'flex', gap: 16, alignItems: 'center' }}>
                    <Icon name={item.icon as IconName} />
                    <span className={sectionItem}>
                      {item.name}
                      {item.new && <Badge variant="pulse" />}
                    </span>
                  </div>
                </ListItem>
              ))}
            </ul>
          </div>
        </Drawer>
      </div>
    )
  }
}
export const Top: Story = {
  args: {
    position: "top"
  },
  render: (args) => {
    const [showDrawer, setShowDrawer] = useState(false)
    const {
      footer,
      header,
      headerButton,
      headerTitle,
      main,
      sectionItem
    } = styles()
    return (
      <div style={{
        overflow: 'hidden', width: '100%', height: '100%'
      }}
      >
        <Button onClick={() => setShowDrawer(!showDrawer)}>Open Drawer</Button>
        <Drawer {...args} open={showDrawer}>
          <div className={header}>
            <Avatar type="icon" />
            <h6 className={headerTitle}>Hello, Design System</h6>
            <div className={headerButton}>
              <IconButton
                onClick={() => setShowDrawer(!showDrawer)}
                ariaLabel="close drawer"
                IconComponent={<Icon name="outlined-navigation-close" />}
              />
            </div>
          </div>
          <Divider />
          <div className={main}>
            <ul style={{ padding: 0, margin: 0, width: '100%' }}>
              {items.body.map((item, id) => (
                <ListItem key={id} interaction="action">
                  <div style={{ display: 'flex', gap: 16, alignItems: 'center' }}>
                    <Icon name={item.icon as IconName} />
                    <span className={sectionItem}>
                      {item.name}
                      {item.new && <Badge variant="pulse" />}
                    </span>
                  </div>
                </ListItem>
              ))}
            </ul>
          </div>
          <div className={footer}>
            <Divider />
            <ul style={{ padding: 0, margin: 0, overflow: 'auto' }}>
              {items.footer.map((item, id) => (
                <ListItem key={id} interaction="action">
                  <div style={{ display: 'flex', gap: 16, alignItems: 'center' }}>
                    <Icon name={item.icon as IconName} />
                    <span className={sectionItem}>
                      {item.name}
                      {item.new && <Badge variant="pulse" />}
                    </span>
                  </div>
                </ListItem>
              ))}
            </ul>
          </div>
        </Drawer>
      </div>
    )
  }
}
export const Bottom: Story = {
  args: {
    position: "bottom"
  },
  render: (args) => {
    const [showDrawer, setShowDrawer] = useState(false)
    const {
      footer,
      header,
      headerButton,
      headerTitle,
      main,
      sectionItem
    } = styles()
    return (
      <div style={{
        overflow: 'hidden', width: '100%', height: '100%'
      }}
      >
        <Button onClick={() => setShowDrawer(!showDrawer)}>Open Drawer</Button>
        <Drawer {...args} open={showDrawer}>
          <div className={header}>
            <Avatar type="icon" />
            <h6 className={headerTitle}>Hello, Design System</h6>
            <div className={headerButton}>
              <IconButton
                onClick={() => setShowDrawer(!showDrawer)}
                ariaLabel="close drawer"
                IconComponent={<Icon name="outlined-navigation-close" />}
              />
            </div>
          </div>
          <Divider />
          <div className={main}>
            <ul style={{ padding: 0, margin: 0, width: '100%' }}>
              {items.body.map((item, id) => (
                <ListItem key={id} interaction="action">
                  <div style={{ display: 'flex', gap: 16, alignItems: 'center' }}>
                    <Icon name={item.icon as IconName} />
                    <span className={sectionItem}>
                      {item.name}
                      {item.new && <Badge variant="pulse" />}
                    </span>
                  </div>
                </ListItem>
              ))}
            </ul>
          </div>
          <div className={footer}>
            <Divider />
            <ul style={{ padding: 0, margin: 0, overflow: 'auto' }}>
              {items.footer.map((item, id) => (
                <ListItem key={id} interaction="action">
                  <div style={{ display: 'flex', gap: 16, alignItems: 'center' }}>
                    <Icon name={item.icon as IconName} />
                    <span className={sectionItem}>
                      {item.name}
                      {item.new && <Badge variant="pulse" />}
                    </span>
                  </div>
                </ListItem>
              ))}
            </ul>
          </div>
        </Drawer>
      </div>
    )
  }
}

const items = {
  body: [
    {
      name: 'Favorites',
      icon: 'outlined-action-love',
      new: false
    },
    {
      name: 'Invite & Earn',
      icon: 'outlined-social-groupofpeople',
      new: false
    },
    {
      name: 'Coupons',
      icon: 'outlined-finance-discount',
      new: true
    },
    {
      name: 'Orders',
      icon: 'outlined-action-request',
      new: false
    },
    {
      name: 'Gifts',
      icon: 'outlined-content-gift',
      new: false
    },
    {
      name: 'Perfumery',
      icon: 'outlined-product-perfumery',
      new: false
    },
    {
      name: 'Body',
      icon: 'outlined-product-dailycare',
      new: false
    },
    {
      name: 'Make Up',
      icon: 'outlined-product-makeup',
      new: false
    },
    {
      name: 'Sales',
      icon: 'outlined-product-promotionproduct',
      new: false
    }
  ],
  footer: [
    {
      name: 'Sign Out',
      icon: 'outlined-navigation-exit',
      new: false
    },
    {
      name: 'Help',
      icon: 'outlined-action-help',
      new: false
    },
    {
      name: 'Privacy Policy',
      icon: 'outlined-social-person',
      new: false
    }
  ]
}

const styles = createUseStyles((theme: Theme) => ({
  sectionItem: {
    fontFamily: [
      theme.navigationDrawer.section.item.primary.fontFamily,
      theme.navigationDrawer.section.item.fallback.fontFamily
    ],
    fontWeight: [
      theme.navigationDrawer.section.item.primary.fontWeight,
      theme.navigationDrawer.section.item.fallback.fontWeight
    ],
    fontSize: theme.navigationDrawer.section.item.fontSize,
    letterSpacing: theme.navigationDrawer.section.item.letterSpacing,
    lineHeight: theme.navigationDrawer.section.item.lineHeight,
    display: 'flex',
    gap: theme.spacing.micro
  },
  footer: {},
  header: {
    padding: theme.spacing.small
  },
  headerButton: {
    display: 'flex',
    justifyContent: 'flex-end',
    padding: 8,
    position: 'absolute',
    top: 0,
    right: 0
  },
  headerTitle: {
    fontFamily: [
      theme.navigationDrawer.header.title.primary.fontFamily,
      theme.navigationDrawer.header.title.fallback.fontFamily
    ],
    fontWeight: theme.navigationDrawer.header.title.primary.fontWeight,
    fontSize: theme.navigationDrawer.header.title.fontSize,
    letterSpacing: theme.navigationDrawer.header.title.letterSpacing,
    lineHeight: theme.navigationDrawer.header.title.lineHeight,
    margin: [theme.spacing.small, 0, theme.spacing.micro]
  },
  main: {
    flex: '1 1 auto',
    display: 'flex',
    overflow: 'hidden',
    position: 'relative',
    overflowY: 'auto'
  }
}))
