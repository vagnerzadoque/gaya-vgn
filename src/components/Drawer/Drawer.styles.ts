/* eslint-disable @typescript-eslint/no-non-null-assertion */

import { createUseStyles } from 'react-jss'
import type { Theme } from '@naturacosmeticos/natds-themes'
import type { DrawerProps, DrawerSize } from './Drawer.props'

const getSize = (size: DrawerSize) => {
  const width = {
    small: 256,
    medium: 375,
    large: 600
  }

  return width[size]
}

const styles = createUseStyles((theme: Theme) => ({
  overlayContainer: ({ open }: DrawerProps) => ({
    backgroundColor: theme.color.highlight,
    display: !open && 'none',
    height: '100%',
    left: 0,
    opacity: theme.opacity.mediumHigh,
    position: 'fixed',
    top: 0,
    width: '100%',
    overflow: 'hidden',
    zIndex: 1199
  }),

  drawer: ({ open = false, size = 'small', position = 'left' }: DrawerProps) => {
    const x = open ? getSize(size) : 0
    const y = open ? getSize(size) : getSize(size) + getSize(size)

    const pos = {
      left: {
        height: '100%',
        left: getSize(size) * -1,
        top: 0,
        transform: `translateX(${x}px)`,
        width: getSize(size)
      },
      right: {
        height: '100%',
        right: getSize(size),
        top: 0,
        transform: `translateX(${y}px)`,
        width: getSize(size)
      },
      top: {
        height: getSize(size),
        left: 0,
        top: getSize(size) * -1,
        transform: `translateY(${x}px)`,
        width: '100%'
      },
      bottom: {
        height: getSize(size),
        left: 0,
        bottom: getSize(size),
        transform: `translateY(${y}px)`,
        width: '100%'
      }
    }
    return {
      display: 'flex',
      flexFlow: 'column nowrap',
      backgroundColor: theme.color.surface,
      color: theme.color.highEmphasis,
      boxShadow: theme.elevation.small,

      ...pos[position],

      justifyContent: 'space-between',
      overflow: 'hidden',
      position: 'absolute',
      transition: '0.3s ease-in-out',
      zIndex: 1200
    }
  }
}))

export default styles
