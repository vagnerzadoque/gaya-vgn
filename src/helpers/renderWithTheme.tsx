import React from 'react'
import { JssProvider, SheetsRegistry } from 'react-jss'
import themes from '@naturacosmeticos/natds-themes'
import { render } from '@testing-library/react'
import { ThemeProvider } from '../ThemeProvider'

const renderWithTheme = (component: JSX.Element) => {
  const sheets = new SheetsRegistry()

  const renderTree = render(
    <JssProvider registry={sheets}>
      <ThemeProvider theme={themes.natura.light}>
        {component}
      </ThemeProvider>
    </JssProvider>
  )

  return { styles: sheets.toString(), component: renderTree }
}

export default renderWithTheme
