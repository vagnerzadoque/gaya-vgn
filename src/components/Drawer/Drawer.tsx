import React from 'react'
import styles from './Drawer.styles'
import { DrawerProps } from './Drawer.props'

const Drawer = React.forwardRef<HTMLDivElement, DrawerProps>(({
  className = '',
  testID,
  ...props
}, ref) => {
  const { overlayContainer, drawer } = styles(props)

  return (
    <>
      <div
        className={`${className} ${drawer} Drawer-gaya`}
        data-testid={testID}
        ref={ref}
        {...props}
      />
      <div className={overlayContainer} onClick={() => props.handleClick?.()} />
    </>
  )
})

export default Drawer
