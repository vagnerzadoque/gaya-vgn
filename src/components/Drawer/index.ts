import Drawer from './Drawer'

export { Drawer }
export { DrawerProps } from './Drawer.props'
export default Drawer
