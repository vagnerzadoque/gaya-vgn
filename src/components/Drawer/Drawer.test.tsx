import * as React from 'react'
import Drawer from '.'
import renderWithTheme from '../../helpers/renderWithTheme'

describe('Drawer component', () => {
  it('should render correctly with default props', () => {
    const { styles, component } = renderWithTheme(<Drawer open>something</Drawer>)

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })

  it('should render correctly with size medium', () => {
    const { styles, component } = renderWithTheme(<Drawer size="medium" open>something</Drawer>)

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })
  it('should render correctly with size large', () => {
    const { styles, component } = renderWithTheme(<Drawer size="large" open>something</Drawer>)

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })
})
