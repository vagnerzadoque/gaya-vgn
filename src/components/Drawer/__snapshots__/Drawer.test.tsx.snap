// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer component should render correctly with default props 1`] = `
[
  ".overlayContainer-0-2-1 {}
.drawer-0-2-2 {}
.overlayContainer-d0-0-2-3 {
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  opacity: 0.56;
  z-index: 1199;
  overflow: hidden;
  position: fixed;
  background-color: #111111;
}
.drawer-d1-0-2-4 {
  top: 0;
  left: -256px;
  color: #363636;
  width: 256px;
  height: 100%;
  display: flex;
  z-index: 1200;
  overflow: hidden;
  position: absolute;
  flex-flow: column nowrap;
  transform: translateX(256px);
  box-shadow: 0 3px 4px 0 rgba(0, 0, 0, 0.14);
  transition: 0.3s ease-in-out;
  justify-content: space-between;
  background-color: #ffffff;
}",
  <div>
    <div
      class=" drawer-0-2-2 drawer-d1-0-2-4 Drawer-gaya"
      open=""
    >
      something
    </div>
    <div
      class="overlayContainer-0-2-1 overlayContainer-d0-0-2-3"
    />
  </div>,
]
`;

exports[`Drawer component should render correctly with size large 1`] = `
[
  ".overlayContainer-0-2-1 {}
.drawer-0-2-2 {}
.overlayContainer-d0-0-2-3 {
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  opacity: 0.56;
  z-index: 1199;
  overflow: hidden;
  position: fixed;
  background-color: #111111;
}
.drawer-d1-0-2-4 {
  top: 0;
  left: -600px;
  color: #363636;
  width: 600px;
  height: 100%;
  display: flex;
  z-index: 1200;
  overflow: hidden;
  position: absolute;
  flex-flow: column nowrap;
  transform: translateX(600px);
  box-shadow: 0 3px 4px 0 rgba(0, 0, 0, 0.14);
  transition: 0.3s ease-in-out;
  justify-content: space-between;
  background-color: #ffffff;
}",
  <div>
    <div
      class=" drawer-0-2-2 drawer-d1-0-2-4 Drawer-gaya"
      open=""
    >
      something
    </div>
    <div
      class="overlayContainer-0-2-1 overlayContainer-d0-0-2-3"
    />
  </div>,
]
`;

exports[`Drawer component should render correctly with size medium 1`] = `
[
  ".overlayContainer-0-2-1 {}
.drawer-0-2-2 {}
.overlayContainer-d0-0-2-3 {
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  opacity: 0.56;
  z-index: 1199;
  overflow: hidden;
  position: fixed;
  background-color: #111111;
}
.drawer-d1-0-2-4 {
  top: 0;
  left: -375px;
  color: #363636;
  width: 375px;
  height: 100%;
  display: flex;
  z-index: 1200;
  overflow: hidden;
  position: absolute;
  flex-flow: column nowrap;
  transform: translateX(375px);
  box-shadow: 0 3px 4px 0 rgba(0, 0, 0, 0.14);
  transition: 0.3s ease-in-out;
  justify-content: space-between;
  background-color: #ffffff;
}",
  <div>
    <div
      class=" drawer-0-2-2 drawer-d1-0-2-4 Drawer-gaya"
      open=""
    >
      something
    </div>
    <div
      class="overlayContainer-0-2-1 overlayContainer-d0-0-2-3"
    />
  </div>,
]
`;
